### PR TITLE
bpo-41513: Improve speed and accuracy of math.hypot()

### PR DESCRIPTION
**Pull request by rhettinger**
https://bugs.python.org/issue41513

### CODE REVIEW

File: Lib/test/test_math.py [hunk 1]
```diff
@@ -795,7 +795,8 @@ def testHypot(self):
         # Verify scaling for extremely large values
         fourthmax = FLOAT_MAX / 4.0
         for n in range(32):
-            self.assertEqual(hypot(*([fourthmax]*n)), fourthmax * math.sqrt(n))
+            self.assertTrue(math.isclose(hypot(*([fourthmax]*n)),
```
Comment (pganssle) [inline Lib/test/test_math.py:798]: Any reason to prefer this over [`assertAlmostEqual`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertAlmostEqual)?

Comment (rhettinger) [reply Lib/test/test_math.py:798]: Yes.  *assertAlmostEqual()* only compares values rounded to seven places.  This is useless for testing very large or very small numbers.  

In contrast, [*math.isclose()*](https://docs.python.org/3/library/math.html#math.isclose) uses both a relative tolerance and absolute tolerance.  The relative tolerance is what we're after in this test.

File: Modules/mathmodule.c [hunk 2]
```diff
@@ -2449,14 +2457,37 @@ vector_norm(Py_ssize_t n, double *vec, double max, int found_nan)
...
     }
+    frexp(max, &max_e);
+    if (max_e >= -1023) {
+        scale = ldexp(1.0, -max_e);
+        assert(max * scale >= 0.5);
+        assert(max * scale < 1.0);
+        for (i=0 ; i < n ; i++) {
+            x = vec[i];
+            assert(Py_IS_FINITE(x) && fabs(x) <= max);
+            x *= scale;
```
Comment (serhiy-storchaka) [inline Modules/mathmodule.c:2473]: What is faster, `x *= scale` or `x = ldexp(x, -max_e)`?

File: Modules/mathmodule.c [hunk 3]
```diff
@@ -2449,14 +2457,37 @@ vector_norm(Py_ssize_t n, double *vec, double max, int found_nan)
     if (max == 0.0 || n <= 1) {
         return max;
     }
+    frexp(max, &max_e);
+    if (max_e >= -1023) {
```
Comment (serhiy-storchaka) [inline Modules/mathmodule.c:2466]: Do we need scaling at all for small absolute `max_e`? E.g. for `-1024/2 < max_e < (1024-64)/2`. Could not it speed up the loop in most common cases?

**Review by serhiy-storchaka — COMMENTED** [review]
If it requires to make tests less strict, it is potentially breaking change, and therefore it should be mentioned in What's New.

File: Modules/mathmodule.c (hunk 2 above)
Comment (rhettinger) [reply Modules/mathmodule.c:2473]: * The test as written was over-specified. It should have been written this way from the start.

* The `x *= scale` is faster than `x = ldexp(x, -max_e)`.  The former is a single, fast in-line instruction and the latter is an external library call.

Here's the generated code for the loop:

```
L284:
    movsd   (%r12,%rax,8), %xmm0
    addq    $1, %rax
    cmpq    %rax, %rbp
    mulsd   %xmm4, %xmm0           <-- x *= scale
    movapd  %xmm0, %xmm1
    mulsd   %xmm0, %xmm1           <-- x *= x
    movapd  %xmm2, %xmm0
    addsd   %xmm1, %xmm2            <-- csum += x
    subsd   %xmm2, %xmm0
    addsd   %xmm1, %xmm0
    addsd   %xmm0, %xmm3
    jg  L284
    subsd   %xmm
```

File: Modules/mathmodule.c (hunk 3 above)
Comment (rhettinger) [reply Modules/mathmodule.c:2466]: Several thoughts:
* Yes, I believe it is needed. 
* The speed effect is minimal. 
* There is no common case for values this small. This is rare.
* Removing the scaling would be a step backwards.

Comment (serhiy-storchaka) [reply Modules/mathmodule.c:2466]: I think that the opposite case, values larger than 1e150 or smaller than 1e-150 (except exact 0) are rare.

Comment (rhettinger) [reply Modules/mathmodule.c:2466]: The scaling isn't just for overflow and underflow.  It is a pre-condition for the accurate summation logic. 

How about we get this code in.  Later you can attempt to further optimize it without losing accuracy, commutativity, and monotonicity.

We've got a straight win here — it's more accurate and faster than what we have now.  I'm happy with that.

Comment (rhettinger) [reply Modules/mathmodule.c:2466]: I updated the comment to indicate why scaling or division by *max* is always desirable, even when not needed for overflow/underflow protection.

**Pull request merged**